From: Alessandro Astone <alessandro.astone@canonical.com>
Date: Thu, 18 Sep 2025 16:09:56 +0200
Subject: xi2: Expose a logical touch device

Create a new GdkDevice used for reporting all touch events from the XI2
Virtual Core Pointer. It tracks the same XI2 pointer device, but reports
to have source GDK_SOURCE_TOUCHSCREEN.

Add a GdkX11SeatXI2 class extending GdkSeatDefault to manage it.

Forwarded: https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/8951
Bug: https://gitlab.gnome.org/GNOME/gtk/-/issues/7515
Bug: https://gitlab.gnome.org/GNOME/gtk/-/issues/7508
Bug: https://gitlab.gnome.org/GNOME/gtk/-/issues/5610
Bug: https://gitlab.gnome.org/GNOME/gtk/-/issues/3542
Bug-Ubuntu: https://launchpad.net/bugs/2119066
Bug-Ubuntu: https://launchpad.net/bugs/2119065
Bug-Ubuntu: https://launchpad.net/bugs/2051886
---
 gdk/gdkseatdefault.c           |  23 +++--
 gdk/gdkseatdefaultprivate.h    |   5 +
 gdk/x11/gdkdevicemanager-xi2.c | 123 +++++++++++++++++++++--
 gdk/x11/gdkseat-xi2-private.h  |  29 ++++++
 gdk/x11/gdkseat-xi2.c          | 217 +++++++++++++++++++++++++++++++++++++++++
 gdk/x11/meson.build            |   1 +
 6 files changed, 383 insertions(+), 15 deletions(-)
 create mode 100644 gdk/x11/gdkseat-xi2-private.h
 create mode 100644 gdk/x11/gdkseat-xi2.c

diff --git a/gdk/gdkseatdefault.c b/gdk/gdkseatdefault.c
index 18d4466..3d86982 100644
--- a/gdk/gdkseatdefault.c
+++ b/gdk/gdkseatdefault.c
@@ -320,11 +320,25 @@ gdk_seat_default_init (GdkSeatDefault *seat)
 {
 }
 
+void
+gdk_seat_default_init_for_logical_pair (GdkSeatDefault *seat,
+                                        GdkDevice *pointer,
+                                        GdkDevice *keyboard)
+{
+  GdkSeatDefaultPrivate *priv;
+
+  priv = gdk_seat_default_get_instance_private (seat);
+  priv->logical_pointer = g_object_ref (pointer);
+  priv->logical_keyboard = g_object_ref (keyboard);
+
+  gdk_seat_device_added (GDK_SEAT (seat), priv->logical_pointer);
+  gdk_seat_device_added (GDK_SEAT (seat), priv->logical_keyboard);
+}
+
 GdkSeat *
 gdk_seat_default_new_for_logical_pair (GdkDevice *pointer,
                                        GdkDevice *keyboard)
 {
-  GdkSeatDefaultPrivate *priv;
   GdkDisplay *display;
   GdkSeat *seat;
 
@@ -334,12 +348,7 @@ gdk_seat_default_new_for_logical_pair (GdkDevice *pointer,
                        "display", display,
                        NULL);
 
-  priv = gdk_seat_default_get_instance_private (GDK_SEAT_DEFAULT (seat));
-  priv->logical_pointer = g_object_ref (pointer);
-  priv->logical_keyboard = g_object_ref (keyboard);
-
-  gdk_seat_device_added (seat, priv->logical_pointer);
-  gdk_seat_device_added (seat, priv->logical_keyboard);
+  gdk_seat_default_init_for_logical_pair (GDK_SEAT_DEFAULT (seat), pointer, keyboard);
 
   return seat;
 }
diff --git a/gdk/gdkseatdefaultprivate.h b/gdk/gdkseatdefaultprivate.h
index 9886d37..b3e68ac 100644
--- a/gdk/gdkseatdefaultprivate.h
+++ b/gdk/gdkseatdefaultprivate.h
@@ -48,6 +48,9 @@ GType     gdk_seat_default_get_type     (void) G_GNUC_CONST;
 
 GdkSeat * gdk_seat_default_new_for_logical_pair         (GdkDevice      *pointer,
                                                          GdkDevice      *keyboard);
+void      gdk_seat_default_init_for_logical_pair        (GdkSeatDefault *seat,
+                                                         GdkDevice      *pointer,
+                                                         GdkDevice      *keyboard);
 
 void      gdk_seat_default_add_physical_device          (GdkSeatDefault *seat,
                                                          GdkDevice      *device);
@@ -58,5 +61,7 @@ void      gdk_seat_default_add_tool                     (GdkSeatDefault *seat,
 void      gdk_seat_default_remove_tool                  (GdkSeatDefault *seat,
                                                          GdkDeviceTool  *tool);
 
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (GdkSeatDefault, g_object_unref)
+
 G_END_DECLS
 
diff --git a/gdk/x11/gdkdevicemanager-xi2.c b/gdk/x11/gdkdevicemanager-xi2.c
index 086995b..ee100d9 100644
--- a/gdk/x11/gdkdevicemanager-xi2.c
+++ b/gdk/x11/gdkdevicemanager-xi2.c
@@ -30,7 +30,7 @@
 #include "gdkdisplay-x11.h"
 #include <glib/gi18n-lib.h>
 #include "gdkkeysyms.h"
-#include "gdkseatdefaultprivate.h"
+#include "gdkseat-xi2-private.h"
 
 #include <X11/Xlib.h>
 #include <X11/Xutil.h>
@@ -456,7 +456,9 @@ is_touchpad_device (GdkDisplay   *display,
 static GdkDevice *
 create_device (GdkX11DeviceManagerXI2 *device_manager,
                GdkDisplay             *display,
-               XIDeviceInfo           *dev)
+               XIDeviceInfo           *dev,
+               const char             *name,
+               int                     source)
 {
   GdkInputSource input_source;
   GdkInputSource touch_source;
@@ -465,7 +467,12 @@ create_device (GdkX11DeviceManagerXI2 *device_manager,
   int num_touches = 0;
   char *vendor_id = NULL, *product_id = NULL;
 
-  if (dev->use == XIMasterKeyboard || dev->use == XISlaveKeyboard)
+  if (!name)
+    name = dev->name;
+
+  if (source > 0)
+    input_source = source;
+  else if (dev->use == XIMasterKeyboard || dev->use == XISlaveKeyboard)
     input_source = GDK_SOURCE_KEYBOARD;
   else if (is_touchpad_device (display, dev))
     input_source = GDK_SOURCE_TOUCHPAD;
@@ -534,7 +541,7 @@ create_device (GdkX11DeviceManagerXI2 *device_manager,
     get_device_ids (display, dev, &vendor_id, &product_id);
 
   device = g_object_new (GDK_TYPE_X11_DEVICE_XI2,
-                         "name", dev->name,
+                         "name", name,
                          "source", input_source,
                          "has-cursor", (dev->use == XIMasterPointer),
                          "display", display,
@@ -577,12 +584,72 @@ ensure_seat_for_device_pair (GdkX11DeviceManagerXI2 *device_manager,
           keyboard = device2;
         }
 
-      seat = gdk_seat_default_new_for_logical_pair (pointer, keyboard);
+      seat = gdk_x11_seat_xi2_new_for_logical_pair (pointer, keyboard);
       gdk_display_add_seat (display, seat);
       g_object_unref (seat);
     }
 }
 
+static void
+ensure_logical_touch_device (gpointer key,
+                             gpointer value,
+                             gpointer user_data)
+{
+  GdkX11DeviceManagerXI2 *device_manager;
+  GdkDevice *physical, *logical_pointer, *logical_touch;
+  GdkDisplay *display;
+  GdkSeat *seat;
+  XIDeviceInfo *info;
+  Display *xdisplay;
+  int logical_pointer_id;
+  int ndevices;
+  g_autofree char *name = NULL;
+
+  device_manager = user_data;
+  physical = g_hash_table_lookup (device_manager->id_table, key);
+
+  if (gdk_device_get_source (physical) != GDK_SOURCE_TOUCHSCREEN)
+    return;
+
+  logical_pointer = g_hash_table_lookup (device_manager->id_table, value);
+  seat = gdk_device_get_seat (logical_pointer);
+  logical_touch = gdk_x11_seat_xi2_get_logical_touch (GDK_X11_SEAT_XI2 (seat));
+
+  if (logical_touch)
+    return;
+
+  display = device_manager->display;
+  xdisplay = GDK_DISPLAY_XDISPLAY (display);
+  logical_pointer_id = _gdk_x11_device_xi2_get_id (GDK_X11_DEVICE_XI2 (logical_pointer));
+  info = XIQueryDevice (xdisplay, logical_pointer_id, &ndevices);
+  g_return_if_fail (ndevices > 0);
+
+  name = g_strdup_printf ("%s touch", gdk_device_get_name (logical_pointer));
+  logical_touch = create_device (device_manager, display, &info[0], name, GDK_SOURCE_TOUCHSCREEN);
+
+  gdk_x11_seat_xi2_set_logical_touch (GDK_X11_SEAT_XI2 (seat), logical_touch);
+
+  device_manager->devices = g_list_append (device_manager->devices, logical_touch);
+
+  XIFreeDeviceInfo (info);
+}
+
+static void
+remove_logical_touch_device (GdkX11DeviceManagerXI2 *device_manager,
+                             GdkDevice              *associated)
+{
+  GdkSeat *seat;
+  GdkDevice *logical_touch;
+
+  seat = gdk_device_get_seat (associated);
+  if (!seat)
+    return;
+
+  logical_touch = gdk_x11_seat_xi2_get_logical_touch (GDK_X11_SEAT_XI2 (seat));
+  device_manager->devices = g_list_remove (device_manager->devices, logical_touch);
+  gdk_x11_seat_xi2_set_logical_touch (GDK_X11_SEAT_XI2 (seat), NULL);
+}
+
 static GdkDevice *
 add_device (GdkX11DeviceManagerXI2 *device_manager,
             XIDeviceInfo           *dev,
@@ -592,7 +659,7 @@ add_device (GdkX11DeviceManagerXI2 *device_manager,
   GdkDevice *device;
 
   display = device_manager->display;
-  device = create_device (device_manager, display, dev);
+  device = create_device (device_manager, display, dev, NULL, -1);
 
   g_hash_table_replace (device_manager->id_table,
                         GINT_TO_POINTER (dev->deviceid),
@@ -618,6 +685,10 @@ add_device (GdkX11DeviceManagerXI2 *device_manager,
 
           seat = gdk_device_get_seat (logical);
           gdk_seat_default_add_physical_device (GDK_SEAT_DEFAULT (seat), device);
+
+          ensure_logical_touch_device (GINT_TO_POINTER (dev->deviceid),
+                                       GINT_TO_POINTER (dev->attachment),
+                                       device_manager);
         }
       else if (dev->use == XIMasterPointer || dev->use == XIMasterKeyboard)
         {
@@ -664,6 +735,7 @@ remove_device (GdkX11DeviceManagerXI2 *device_manager,
 
   if (device)
     {
+      remove_logical_touch_device (device_manager, device);
       detach_from_seat (device);
 
       g_hash_table_remove (device_manager->id_table,
@@ -770,6 +842,7 @@ gdk_x11_device_manager_xi2_constructed (GObject *object)
   g_hash_table_destroy (logical_devices);
 
   g_hash_table_foreach (physical_devices, relate_physical_devices, object);
+  g_hash_table_foreach (physical_devices, ensure_logical_touch_device, object);
   g_hash_table_destroy (physical_devices);
 
   /* Connect to hierarchy change events */
@@ -937,6 +1010,7 @@ handle_hierarchy_changed (GdkX11DeviceManagerXI2 *device_manager,
 
                   seat = gdk_device_get_seat (logical);
                   gdk_seat_default_add_physical_device (GDK_SEAT_DEFAULT (seat), physical);
+                  ensure_logical_touch_device (GINT_TO_POINTER (ev->info[i].deviceid), logical, device_manager);
                 }
             }
         }
@@ -1221,11 +1295,12 @@ get_event_surface (GdkEventTranslator *translator,
                   XIEvent            *ev,
                   GdkSurface         **surface_p)
 {
+  GdkX11DeviceManagerXI2 *device_manager = GDK_X11_DEVICE_MANAGER_XI2 (translator);
   GdkDisplay *display;
   GdkSurface *surface = NULL;
   gboolean should_have_window = TRUE;
 
-  display = GDK_X11_DEVICE_MANAGER_XI2 (translator)->display;
+  display = device_manager->display;
 
   switch (ev->evtype)
     {
@@ -1249,11 +1324,21 @@ get_event_surface (GdkEventTranslator *translator,
           {
             GdkDeviceGrabInfo *info;
             GdkDevice *device;
+            GdkSeat *seat;
             gulong serial;
 
-            device = g_hash_table_lookup (GDK_X11_DEVICE_MANAGER_XI2 (translator)->id_table,
+            device = g_hash_table_lookup (device_manager->id_table,
                                           GUINT_TO_POINTER (((XIDeviceEvent *) ev)->deviceid));
 
+            if ((ev->evtype == XI_TouchUpdate ||
+                 ev->evtype == XI_TouchBegin ||
+                 ev->evtype == XI_TouchEnd) &&
+                gdk_device_get_source (device) == GDK_SOURCE_TOUCHSCREEN)
+              {
+                seat = gdk_device_get_seat (device);
+                device = gdk_x11_seat_xi2_get_logical_touch (GDK_X11_SEAT_XI2 (seat));
+              }
+
             serial = _gdk_display_get_next_serial (display);
             info = _gdk_display_has_device_grab (display, device, serial);
 
@@ -1784,6 +1869,16 @@ gdk_x11_device_manager_xi2_translate_event (GdkEventTranslator *translator,
             break;
           }
 
+#ifdef XINPUT_2_2
+        if (xev->flags & XIPointerEmulated &&
+            gdk_device_get_source (source_device) == GDK_SOURCE_TOUCHSCREEN)
+          {
+            /* Touch drag&drop */
+            GdkSeat *seat = gdk_device_get_seat (device);
+            device = gdk_x11_seat_xi2_get_logical_touch (GDK_X11_SEAT_XI2 (seat));
+          }
+#endif
+
         axes = translate_axes (device,
                                (double) xev->event_x / scale,
                                (double) xev->event_y / scale,
@@ -1827,6 +1922,12 @@ gdk_x11_device_manager_xi2_translate_event (GdkEventTranslator *translator,
         source_device = g_hash_table_lookup (device_manager->id_table,
                                              GUINT_TO_POINTER (xev->sourceid));
 
+        if (G_LIKELY (gdk_device_get_source (source_device) == GDK_SOURCE_TOUCHSCREEN))
+          {
+            GdkSeat *seat = gdk_device_get_seat (device);
+            device = gdk_x11_seat_xi2_get_logical_touch (GDK_X11_SEAT_XI2 (seat));
+          }
+
         state = _gdk_x11_device_xi2_translate_state (&xev->mods, &xev->buttons, &xev->group);
         if (ev->evtype == XI_TouchBegin)
           state |= GDK_BUTTON1_MASK;
@@ -1877,6 +1978,12 @@ gdk_x11_device_manager_xi2_translate_event (GdkEventTranslator *translator,
         source_device = g_hash_table_lookup (device_manager->id_table,
                                              GUINT_TO_POINTER (xev->sourceid));
 
+        if (G_LIKELY (gdk_device_get_source (source_device) == GDK_SOURCE_TOUCHSCREEN))
+          {
+            GdkSeat *seat = gdk_device_get_seat (device);
+            device = gdk_x11_seat_xi2_get_logical_touch (GDK_X11_SEAT_XI2 (seat));
+          }
+
         state = _gdk_x11_device_xi2_translate_state (&xev->mods, &xev->buttons, &xev->group);
         state |= GDK_BUTTON1_MASK;
 
diff --git a/gdk/x11/gdkseat-xi2-private.h b/gdk/x11/gdkseat-xi2-private.h
new file mode 100644
index 0000000..4a5c23f
--- /dev/null
+++ b/gdk/x11/gdkseat-xi2-private.h
@@ -0,0 +1,29 @@
+/* GDK - The GIMP Drawing Kit
+ * Copyright (C) 2025 Canonical Ltd.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Alessandro Astone <alessandro.astone@canonical.com>
+ */
+
+#include "gdkseatdefaultprivate.h"
+
+#define GDK_TYPE_X11_SEAT_XI2 (gdk_x11_seat_xi2_get_type ())
+G_DECLARE_FINAL_TYPE (GdkX11SeatXI2, gdk_x11_seat_xi2, GDK, X11_SEAT_XI2, GdkSeatDefault)
+
+GdkSeat *gdk_x11_seat_xi2_new_for_logical_pair (GdkDevice *pointer,
+                                                GdkDevice *keyboard);
+
+GdkDevice *gdk_x11_seat_xi2_get_logical_touch (GdkX11SeatXI2 *seat);
+void       gdk_x11_seat_xi2_set_logical_touch (GdkX11SeatXI2 *seat, GdkDevice *device);
diff --git a/gdk/x11/gdkseat-xi2.c b/gdk/x11/gdkseat-xi2.c
new file mode 100644
index 0000000..74a79c0
--- /dev/null
+++ b/gdk/x11/gdkseat-xi2.c
@@ -0,0 +1,217 @@
+/* GDK - The GIMP Drawing Kit
+ * Copyright (C) 2025 Canonical Ltd.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Alessandro Astone <alessandro.astone@canonical.com>
+ */
+
+#include "gdkdisplayprivate.h"
+#include "gdkseat-xi2-private.h"
+
+struct _GdkX11SeatXI2
+{
+  GdkSeatDefault parent_instance;
+
+  GdkDevice *logical_touch;
+};
+
+G_DEFINE_TYPE (GdkX11SeatXI2, gdk_x11_seat_xi2, GDK_TYPE_SEAT_DEFAULT)
+
+static void gdk_x11_seat_xi2_finalize (GObject *object);
+static GdkDevice *gdk_x11_seat_xi2_get_logical_device (GdkSeat             *seat,
+                                                       GdkSeatCapabilities  capability);
+static GList *gdk_x11_seat_xi2_get_devices (GdkSeat             *seat,
+                                            GdkSeatCapabilities  capabilities);
+static GdkGrabStatus gdk_x11_seat_xi2_grab (GdkSeat                *seat,
+                                            GdkSurface              *surface,
+                                            GdkSeatCapabilities     capabilities,
+                                            gboolean                owner_events,
+                                            GdkCursor              *cursor,
+                                            GdkEvent               *event,
+                                            GdkSeatGrabPrepareFunc  prepare_func,
+                                            gpointer                prepare_func_data);
+static void gdk_x11_seat_xi2_ungrab (GdkSeat *seat);
+
+static void
+gdk_x11_seat_xi2_class_init (GdkX11SeatXI2Class *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+  GdkSeatClass *seat_class = GDK_SEAT_CLASS (klass);
+
+  object_class->finalize = gdk_x11_seat_xi2_finalize;
+
+  seat_class->get_logical_device = gdk_x11_seat_xi2_get_logical_device;
+  seat_class->get_devices = gdk_x11_seat_xi2_get_devices;
+  seat_class->grab = gdk_x11_seat_xi2_grab;
+  seat_class->ungrab = gdk_x11_seat_xi2_ungrab;
+}
+
+static void
+gdk_x11_seat_xi2_init (GdkX11SeatXI2 *seat)
+{
+}
+
+static void
+gdk_x11_seat_xi2_finalize (GObject *object)
+{
+  GdkX11SeatXI2 *seat = GDK_X11_SEAT_XI2 (object);
+  g_clear_object (&seat->logical_touch);
+
+  G_OBJECT_CLASS (gdk_x11_seat_xi2_parent_class)->finalize (object);
+}
+
+GdkSeat *
+gdk_x11_seat_xi2_new_for_logical_pair (GdkDevice *pointer,
+                                       GdkDevice *keyboard)
+{
+  GdkDisplay *display;
+  GdkSeat *seat;
+
+  display = gdk_device_get_display (pointer);
+
+  seat = g_object_new (GDK_TYPE_X11_SEAT_XI2,
+                       "display", display,
+                       NULL);
+
+  gdk_seat_default_init_for_logical_pair (GDK_SEAT_DEFAULT (seat), pointer, keyboard);
+
+  return seat;
+}
+
+GdkDevice *
+gdk_x11_seat_xi2_get_logical_touch (GdkX11SeatXI2 *seat)
+{
+  return seat->logical_touch;
+}
+
+void
+gdk_x11_seat_xi2_set_logical_touch (GdkX11SeatXI2 *seat, GdkDevice *device)
+{
+  GdkDevice *logical_pointer;
+
+  g_clear_object (&seat->logical_touch);
+  if (!device)
+    return;
+
+  seat->logical_touch = g_object_ref (device);
+  logical_pointer = gdk_seat_get_pointer (GDK_SEAT (seat));
+
+  _gdk_device_set_associated_device (seat->logical_touch, logical_pointer);
+  gdk_seat_device_added (GDK_SEAT (seat), seat->logical_touch);
+}
+
+static GdkDevice *
+gdk_x11_seat_xi2_get_logical_device (GdkSeat             *seat,
+                                     GdkSeatCapabilities  capability)
+{
+  GdkX11SeatXI2 *seat_xi2 = GDK_X11_SEAT_XI2 (seat);
+
+  if (capability == GDK_SEAT_CAPABILITY_TOUCH && seat_xi2->logical_touch)
+    return seat_xi2->logical_touch;
+
+  return GDK_SEAT_CLASS (gdk_x11_seat_xi2_parent_class)->get_logical_device (seat, capability);
+}
+
+static GList *
+gdk_x11_seat_xi2_get_devices (GdkSeat             *seat,
+                              GdkSeatCapabilities  capabilities)
+{
+  GdkX11SeatXI2 *seat_xi2 = GDK_X11_SEAT_XI2 (seat);
+  GList *devices;
+
+  devices = GDK_SEAT_CLASS (gdk_x11_seat_xi2_parent_class)->get_devices (seat, capabilities);
+
+  if (capabilities & GDK_SEAT_CAPABILITY_TOUCH && seat_xi2->logical_touch)
+    devices = g_list_prepend (devices, seat_xi2->logical_touch);
+
+  return devices;
+}
+
+static GdkGrabStatus
+gdk_x11_seat_xi2_grab (GdkSeat                *seat,
+                       GdkSurface              *surface,
+                       GdkSeatCapabilities     capabilities,
+                       gboolean                owner_events,
+                       GdkCursor              *cursor,
+                       GdkEvent               *event,
+                       GdkSeatGrabPrepareFunc  prepare_func,
+                       gpointer                prepare_func_data)
+{
+  GdkGrabStatus status;
+
+  /* Grab keyboard and pointer.
+   * Strip the TOUCH capability to force emulated pointer events
+   */
+  status = GDK_SEAT_CLASS (gdk_x11_seat_xi2_parent_class)->grab (seat,
+                                                                 surface,
+                                                                 capabilities & ~GDK_SEAT_CAPABILITY_TOUCH,
+                                                                 owner_events,
+                                                                 cursor,
+                                                                 event,
+                                                                 prepare_func,
+                                                                 prepare_func_data);
+
+  if (status != GDK_GRAB_SUCCESS)
+    return status;
+
+  /* Add a gdk device grab on the logical touch device */
+  if (capabilities & GDK_SEAT_CAPABILITY_TOUCH)
+    {
+      GdkX11SeatXI2 *seat_xi2 = GDK_X11_SEAT_XI2 (seat);
+      GdkDisplay *display;
+      gulong serial;
+      guint32 evtime;
+
+      display = gdk_surface_get_display (surface);
+      serial = _gdk_display_get_next_serial (display);
+      evtime = event ? gdk_event_get_time (event) : GDK_CURRENT_TIME;
+
+      _gdk_display_add_device_grab (display,
+                                    seat_xi2->logical_touch,
+                                    surface,
+                                    owner_events,
+                                    GDK_TOUCH_MASK,
+                                    serial,
+                                    evtime,
+                                    FALSE);
+    }
+
+  return status;
+}
+
+void
+gdk_x11_seat_xi2_ungrab (GdkSeat *seat)
+{
+  GdkX11SeatXI2 *seat_xi2 = GDK_X11_SEAT_XI2 (seat);
+
+  GDK_SEAT_CLASS (gdk_x11_seat_xi2_parent_class)->ungrab (seat);
+
+  if (seat_xi2->logical_touch)
+    {
+      GdkDeviceGrabInfo *touch_grab;
+      GdkDisplay *display;
+
+      display = gdk_device_get_display (seat_xi2->logical_touch);
+      touch_grab = _gdk_display_get_last_device_grab (display, seat_xi2->logical_touch);
+
+      if (touch_grab)
+        {
+          gulong serial = _gdk_display_get_next_serial (display);
+
+          touch_grab->serial_end = serial;
+          _gdk_display_device_grab_update (display, seat_xi2->logical_touch, serial);
+        }
+    }
+}
diff --git a/gdk/x11/meson.build b/gdk/x11/meson.build
index c9aa538..d876757 100644
--- a/gdk/x11/meson.build
+++ b/gdk/x11/meson.build
@@ -16,6 +16,7 @@ gdk_x11_public_sources = files([
   'gdkproperty-x11.c',
   'gdkscreen-x11.c',
   'gdkselection-x11.c',
+  'gdkseat-xi2.c',
   'gdksurface-x11.c',
   'gdkxid.c',
   'xsettings-client.c',
